### PR TITLE
Changing Quarkus transaction handling for JPA map storage to JTA

### LINKED ISSUE #13222: Followup: Revisit JTA vs. RESOURCE_LOCAL for JPA map storage for Quarkus and other Setups
### Description

When implementing the JPA Map store in Quarkus, in issue #13200 it was found that the autocommit for the DB connection wasn't disabled.

In the Undertow setup for JPA Map storage, Hibernate disabled the autocommit. In the Quarkus setup with the legacy store this is disabled by the XA transaction manger. 

A workaround in the `QuarkusJpaMapStorageProviderFactory` disables the autocommit. 

A better solution would be to disable it in the Agroal connection pool, as that will reset the autocommit on every returned connection, which seems to be overhead. 

Unfortunately I didn't find an option to change the behavior of any component involved, so there is a workaround in place.

### Discussion

_No response_

### Motivation

Avoid switching the autocommit manually back and forth when retrieving and returning the connection. Ideally it should be handled in the connection pool. 

This is also a solution advocated for here: https://vladmihalcea.com/why-you-should-always-use-hibernate-connection-provider_disables_autocommit-for-resource-local-jpa-transactions/

### Details

_No response_

### PR DESCRIPTION
**Pull request by ahus1**
This has been recommended as the supported way of transaction handling by the Quarkus team.

Closes #13222

### CODE REVIEW

**Comment by ahus1** [comment]
@vramik / @pedroigor - this is my first attempt to change the transaction handling on Quarkus to use JTA. I've adding you as reviewers for Quarkus (Pedro) and for JPA (Vlasta) as we've been discussion this previously. Please let me know if you don't find the time. 

Pinging @sguilhen if you want to have a look as well optionally - please add yourself as a review if you think this makes sense.

I found that it is best to use a separate JTA transaction for Liquibase updates, as once I unwrapped the DB connection some data didn't end up in the database as it should. Unwrapping of the DB connection was necessary as on a PostgreSQL connection one would need to run rollback to re-use the connection after an SQL error. And as there are temporary SQL errors when creating the first `DATABASECHANGELOG` and `DATABASECHANGELOGLOCK` tables, unwrapping was the way to go.

After switching to JTA, I didn't see the need to switch to XA at the same time, do you think otherwise?

IntelliJ gave me a warning when throwing an exception in the finally block - happy to hear if someone has a different idea on how to handle this.

https://github.com/keycloak/keycloak/blob/f0a67ff5753858c2f5c241995f9c3771031bad34/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java#L432-L440

**Review by sguilhen — COMMENTED** [review]
@ahus1 Aside from the failing tests, which might or might not be related, I have to ask if you had a chance to test these changes with CockroachDB. I recall something about CRDB + JTA being an issue back when a community member was trying to get the legacy store to run with CRDB.

**Comment by ahus1** [comment]
@sguilhen - thanks for bringing up CockroachDB. 

The good news: It passed all the smoke tests locally: I was able to spin it up starting with an empty database, doing all migrations initializing the database, and then to log in to the admin console and browse the entities.  So I'd say there is no blocker in sight for CockroachDB. 

I'll do more exploratory testing to see if something hides behind a tree, and also set up some GHA run for CockroachDB that I'll enable in a branch.

**Comment by sguilhen** [comment]
> @sguilhen - thanks for bringing up CockroachDB.
> 
> The good news: It passed all the smoke tests locally: I was able to spin it up starting with an empty database, doing all migrations initializing the database, and then to log in to the admin console and browse the entities. So I'd say there is no blocker in sight for CockroachDB.
> 
> I'll do more exploratory testing to see if something hides behind a tree, and also set up some GHA run for CockroachDB that I'll enable in a branch.

Great, thank you @ahus1 ! I think this initial testing looks very promising because IIRC the legacy store would fail when processing the changesets, so the fact that this is working here looks good.

File: model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java [hunk 1]
```diff
@@ -260,7 +271,7 @@ public MapStorageProvider create(KeycloakSession session) {
         // check the session for a cached provider before creating a new one.
         JpaMapStorageProvider provider = session.getAttribute(this.sessionProviderKey, JpaMapStorageProvider.class);
         if (provider == null) {
-            provider = new JpaMapStorageProvider(this, session, getEntityManager(), this.sessionTxKey);
+            provider = new JpaMapStorageProvider(this, session, PersistenceExceptionConverter.create(session, getEntityManager()), this.sessionTxKey, this.jtaEnabled);
```
[the remark this answers is not in the document]
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java:274]: The instrumentation is done for the entity manager only, and an auto-flush could still be caused when executing a query, which are not instrumented. The alternative would be to wrap also the queries returned by the entity manager. 

On another note, I was wondering if this should be handled differently anyway...

[the remark this answers is not in the document]
Comment (sguilhen) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java:274]: Makes sense, thanks! Any suggestions on how to handle this differently?

[the remark this answers is not in the document]
Comment (ahus1) [reply model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java:274]: I was thinking to skip the instrumentation of the entity manager altogether and wrap only each method in the JPA transaction class, which should (?) have the same effect. The the scope of this issue, I'd keep it as is.

**Comment by ahus1** [comment]
Some tests fail with 

> Server returned HTTP response code: 429 for URL: https://bitbucket.org/ariya/phantomjs/downloads/phantomjs-2.1.1-linux-x86_64.tar.bz2

where 429 means Too Many Requests.

Will do a rerun once all running builds succeeded. 

If you want to do a review of the different steps while it is not being squashed, now would be a good time :-)

**Comment by pedroigor** [comment]
@ahus1 For me, whatever works best for the JPA map store works. As we discussed before, Quarkus recommends JTA anyways.

IIRC, by not enabling XA we are basically using Nayarana TM but without two-phase commit support from the underling datasource.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/transaction/QuarkusJtaTransactionManagerLookup.java [hunk 2]
```diff
@@ -70,6 +69,6 @@ public int order() {
 
     @Override
     public boolean isSupported() {
-        return !Profile.isFeatureEnabled(Profile.Feature.MAP_STORAGE);
+        return true;
```
Comment (vramik) [inline quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/transaction/QuarkusJtaTransactionManagerLookup.java:72]: I'm just wondering if it makes sense to implement `EnvironmentDependentProviderFactory` after all when it becomes supported always.

**Review by vramik — COMMENTED** [review]
Thank you @ahus1! It looks great to me. I have just one minor question.

File: quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/transaction/QuarkusJtaTransactionManagerLookup.java (hunk 2 above)
Comment (ahus1) [reply quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/transaction/QuarkusJtaTransactionManagerLookup.java:72]: Thanks for noting this, I forgot about that part. In the latest commit this has now been cleaned up.

**Comment by ahus1** [comment]
This is now ready for a final review when your time permits (no earlier than Monday). 

As agreed in our team meeting, the GHA for CockroachDB are not part of this PR. The preliminary change to the GHA was executed here in my repo: https://github.com/ahus1/keycloak/actions/runs/3647760371/jobs/6160550528

All tests passed on CockroachDB. The other failure of a Quarkus test in another build was due to a problem in the nip.io DNS AFAIK.

**Review by sguilhen — APPROVED** [review]
Given @pedroigor 's positive review, and the verification that everything works with CockroachDB, I'm approving this.

**Review by pedroigor — APPROVED** [review]

**Pull request merged**